File: services/llm-gateway/src/schemas.py
from __future__ import annotations

from dataclasses import asdict, dataclass, field
from typing import Any


@dataclass(frozen=True)
class EvidenceChunk:
    chunk_id: str
    doc_id: str
    title: str
    content: str
    visibility: str
    data_level: str
    allowed_roles: list[str]
    source_uri: str

    @classmethod
    def from_chunk(cls, chunk: dict[str, Any]) -> "EvidenceChunk":
        return cls(
            chunk_id=str(chunk["chunk_id"]),
            doc_id=str(chunk["doc_id"]),
            title=str(chunk["title"]),
            content=str(chunk["content"]),
            visibility=str(chunk["visibility"]),
            data_level=str(chunk["data_level"]),
            allowed_roles=list(chunk.get("allowed_roles", [])),
            source_uri=str(chunk.get("source_uri", "")),
        )

    def to_prompt_dict(self) -> dict[str, Any]:
        return asdict(self)


@dataclass(frozen=True)
class LLMRequest:
    user_role: str
    intent: str
    user_query: str
    allowed_evidence: list[EvidenceChunk]
    answer_policy: dict[str, Any] = field(default_factory=dict)
    output_format: str = "plain_text_with_sources"
    risk_level: str = "low"
    session_id: str = ""
    campus: str = "all"

    def to_policy_dict(self) -> dict[str, Any]:
        return {
            "task": "admissions_answer",
            "message": self.user_query,
            "intent": self.intent,
            "scope": {
                "role": self.user_role,
                "campus": self.campus,
            },
            "risk_level": self.risk_level,
            "session_id": self.session_id,
            "answer_policy": self.answer_policy,
            "output_format": self.output_format,
            "evidence": [chunk.to_prompt_dict() for chunk in self.allowed_evidence],
        }

File: services/llm-gateway/ts-src/schemas.ts
import { z } from "zod";

// --- str() helper matching Python's str() for None handling ---
function pyStr(v: unknown): string {
  if (v === null) return "None";
  return String(v);
}

// Matches Python's `list(chunk.get(key, default))`
function listFromChunk(
  chunk: Record<string, unknown>,
  key: string,
  defaultVal: string[],
): string[] {
  if (!(key in chunk)) return defaultVal;
  const val = chunk[key];
  if (val === null) throw new TypeError("'NoneType' object is not iterable");
  if (Array.isArray(val)) return val.map(pyStr);
  return [pyStr(val)];
}

// Matches Python's `str(chunk.get(key, default))`
function strFromChunk(
  chunk: Record<string, unknown>,
  key: string,
  defaultVal: string,
): string {
  if (!(key in chunk)) return defaultVal;
  return pyStr(chunk[key]);
}

// --- EvidenceChunk ---

export const EvidenceChunkSchema = z.object({
  chunk_id: z.string(),
  doc_id: z.string(),
  title: z.string(),
  content: z.string(),
  visibility: z.string(),
  data_level: z.string(),
  allowed_roles: z.array(z.string()),
  source_uri: z.string(),
}).strict();

export type EvidenceChunk = z.infer<typeof EvidenceChunkSchema>;

export function evidenceChunkFromChunk(
  chunk: Record<string, unknown>,
): EvidenceChunk {
  function require(key: string): unknown {
    if (!(key in chunk)) throw new Error(`KeyError: '${key}'`);
    return chunk[key];
  }
  return EvidenceChunkSchema.parse({
    chunk_id: pyStr(require("chunk_id")),
    doc_id: pyStr(require("doc_id")),
    title: pyStr(require("title")),
    content: pyStr(require("content")),
    visibility: pyStr(require("visibility")),
    data_level: pyStr(require("data_level")),
    allowed_roles: listFromChunk(chunk, "allowed_roles", []),
    source_uri: strFromChunk(chunk, "source_uri", ""),
  });
}

// --- LLMRequest ---

export const LLMRequestSchema = z.object({
  user_role: z.string(),
  intent: z.string(),
  user_query: z.string(),
  allowed_evidence: z.array(EvidenceChunkSchema),
  answer_policy: z.record(z.string(), z.unknown()).default({}),
  output_format: z.string().default("plain_text_with_sources"),
  risk_level: z.string().default("low"),
  session_id: z.string().default(""),
  campus: z.string().default("all"),
}).strict();

export type LLMRequest = z.infer<typeof LLMRequestSchema>;

export function llmRequestToPolicyDict(
  req: LLMRequest,
): Record<string, unknown> {
  return {
    task: "admissions_answer",
    message: req.user_query,
    intent: req.intent,
    scope: {
      role: req.user_role,
      campus: req.campus,
    },
    risk_level: req.risk_level,
    session_id: req.session_id,
    answer_policy: req.answer_policy,
    output_format: req.output_format,
    evidence: req.allowed_evidence,
  };
}
